### PR TITLE
Implement the createTown game command

### PR DESCRIPTION
**Pull request by AaronVanGeffen**

### CODE REVIEW

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                    if (dx < 10 && gameState.rng.randNext() & 0xFF)
+                    {
+                        continue;
+                    }
+                }
+
+                // loc_496E89
+                for (auto& town : TownManager::towns())
+                {
+                    auto xDiff = std::abs(pos.x - town.x);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:87]: We have the helper Manhattan distance function for this

Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:87]: We have the helper Manhattan distance function for this

**Comment by AaronVanGeffen** [comment]
Testing this yields my `Invalid position -- trying again` debug statements in pretty much all cases. I'm probably not using the `randNext()` function correctly. Please advise.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        Pos2 pos = args.pos;
+        Town* newTown = nullptr;
+
+        if (args.pos.x == -1)
+        {
+            bool foundPos = false;
+            for (auto attempts = 200; attempts > 0; attempts--)
+            {
+                uint32_t rand = gameState.rng.randNext();
+                pos = Pos2(((rand & 0xFFFF) * kMapRows) << 5UL, ((rand >> 16) * kMapColumns) << 5UL);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:99]: This line leads to very wrong map positions being generated. Not sure how to improve/fix it yet.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        for (int i = 8; i > 0; i--)
+        {
+            for (int j = growthFactor; j > 0; j--)
+            {
+                newTown->grow(0xFF);
+                newTown->recalculateSize();
+            }
+
+            setCurrentYear(currentYear + 7);
+        }
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:187]: Not sure about this. It feels like too much growth.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        Pos2 pos = args.pos;
+        Town* newTown = nullptr;
+
+        if (args.pos.x == -1)
+        {
+            bool foundPos = false;
+            for (auto attempts = 200; attempts > 0; attempts--)
+            {
+                uint32_t rand = gameState.rng.randNext();
+                pos = Pos2((((rand >> 16) * kMapColumns) >> 16) << 5, (((rand & 0xFFFF) * kMapRows) >> 16) << 5);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:100]:
```suggestion
                pos = Pos2((((rand >> 16) * kMapColumns) >> 16) * kTileSize, (((rand & 0xFFFF) * kMapRows) >> 16) * kTileSize);
```

Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp:100]: Or just convert from TilePos2 to Pos2 using `toWorldSpace`. Probably more readable than further fiddling with constants as well.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp (hunk 3 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp:178]: I think these kinds of buildings shouldn't be built in-game during 1900 though. Have to check with master.

Unnamed

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                newTown->grow(0xFF);
+                newTown->recalculateSize();
+            }
+
+            setCurrentYear(currentYear + 7);
+        }
+
+        setCurrentYear(currentYear);
+
+        newTown->history[newTown->historySize - 1] = newTown->population / 50;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:182]: missing a `std::max(newTown->population / 50, 255)`

Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp:182]:
```suggestion
        newTown->history[newTown->historySize - 1] = std::max(newTown->population / 50, 255)
```

File: src/OpenLoco/src/GameCommands/CreateTown.cpp (hunk 3 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp]: Found and fixed the mistake.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    static uint16_t sub_4C5604(Pos2 pos)
+    {
+        registers regs;
+        regs.ax = pos.x;
+        regs.cx = pos.y;
+        call(0x004C5604, regs);
+        return regs.dx;
+    }
+
+    static bool checkSurroundings(Pos2 pos)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:51]: add the address comment // 0x00496E09

Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp:51]:
```suggestion
    // 0x00496E09
    static bool checkSurroundings(Pos2 pos)
```

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+        if (regs.esi != -1)
+            return reinterpret_cast<Town*>(regs.esi);
+        else
+            return nullptr;
+    }
+
+    // 0x004C5604
+    // TODO: used by other functions; move elsewhere
+    static uint16_t sub_4C5604(Pos2 pos)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:42]: countNearbyWaterTiles

Comment (AaronVanGeffen) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp]: Haha, I'd just renamed it to `countNearbyWaterTiles` as per your earlier suggestion. Saw that there was a `countSurroundingWaterTiles` already in the TileManager, indeed.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            }
+        }
+        else
+        {
+            bool foundPos = false;
+            for (int attempts = 0; attempts < 40; attempts++)
+            {
+                if (attempts == 0)
+                {
+                    uint32_t rand = gameState.rng.randNext();
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:113]: this doesn't look correct. attempt 0 is a just use the current x,y. Attempts 1 -> 10 is the & 0x60. And attempt 11 > 39 is the & 0xE0.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                    pos += Pos2(((rand >> 4) & 0xE0) - 0x60, (rand & 0xE0) - 0x60);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:114]: None of these calcs look correct. Nothing should be modifying pos. This should be a new variable.

Comment (duncanspumpkin) [reply src/OpenLoco/src/GameCommands/CreateTown.cpp:114]: Its basically doing a random number between -3 -> 4 added onto pos x and y.
```cpp
TilePos2 randOffset(((rand >> 9) & 7) - 3, ((rand >> 5) & 7) - 3); // Random value -3 -> 4 x and y
const auto attemptLoc = pos + World::toWorldSpace(randOffset);
```

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                }
+                else if (attempts <= 10)
+                {
+                    uint32_t rand = gameState.rng.randNext();
+                    pos += Pos2(((rand >> 4) & 0x60) - kTileSize, (rand & 0x60) - kTileSize);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:119]:
```cpp
TilePos2 randOffset(((rand >> 9) & 3) - 2, ((rand >> 5) & 3) - 2); // Random value -2 -> 1 x and y
const auto attemptLoc = pos + World::toWorldSpace(randOffset);
```

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+                }
+
+                if (pos.x < 160 || pos.y < 160 || pos.x > 12128 || pos.y > 12128)
+                {
+                    continue;
+                }
+
+                if (checkSurroundings(pos))
+                {
+                    foundPos = true;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:129]: If the above changes are made you'll need to copy the attemptLoc into pos at this point.

File: src/OpenLoco/src/GameCommands/CreateTown.cpp [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+                return FAILURE;
+            }
+        }
+        else
+        {
+            bool foundPos = false;
+            for (int attempts = 0; attempts < 40; attempts++)
+            {
+                Pos2 attemptPos = pos;
+                if (attempts <= 10)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/CreateTown.cpp:116]: 0 is <= 10. Think you'll need a `if (attempts == 0){ // do nothing attemptPos = pos }`

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**